### ISSUE TITLE
More Gracefully Handle Insufficient Disk Space

### ISSUE THREAD

**Issue by zedseven** [keuin/KBackup-Fabric#3]
I've been running this on my server for a while now and it works really well, but today when I ran a backup the backup failed because the server didn't have enough disk space (it's a small server with like 50GB of storage).

This in itself is fine - it's easy enough to make more space, but the operation threw an exception and led to a loss of my player advancements data.

Here is the relevant portion of the log:
```
...
[18:07:05] [Server thread/INFO]: [KBackup] Start backup...
[18:07:05] [Server thread/INFO]: Making backup, please wait ...
[18:07:07] [Server thread/INFO]: ThreadedAnvilChunkStorage (ConglomerateWorld3): All chunks are saved
[18:07:07] [Server thread/INFO]: ThreadedAnvilChunkStorage (DIM-1): All chunks are saved
[18:07:07] [Server thread/INFO]: ThreadedAnvilChunkStorage (DIM1): All chunks are saved
[18:07:07] [BackupWorker/INFO]: [KBackup] zip(srcPath=/srv/multi-host/minecraft/fabric/1.16.5/ConglomerateWorld3, destPath=/srv/multi-host/minecraft/fabric/1.16.5/backups)
[18:07:07] [BackupWorker/INFO]: [KBackup] Compressing level ...
[18:08:37] [BackupWorker/INFO]: [STDERR]: java.io.IOException: No space left on device
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at java.base/java.io.FileOutputStream.writeBytes(Native Method)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at java.base/java.io.FileOutputStream.write(FileOutputStream.java:354)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at java.base/java.util.zip.CheckedOutputStream.write(CheckedOutputStream.java:74)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at java.base/java.util.zip.DeflaterOutputStream.deflate(DeflaterOutputStream.java:253)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at java.base/java.util.zip.ZipOutputStream.closeEntry(ZipOutputStream.java:256)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at java.base/java.util.zip.ZipOutputStream.finish(ZipOutputStream.java:361)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at java.base/java.util.zip.DeflaterOutputStream.close(DeflaterOutputStream.java:238)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at java.base/java.util.zip.ZipOutputStream.close(ZipOutputStream.java:378)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at com.keuin.kbackupfabric.util.ZipUtil.makeBackupZip(ZipUtil.java:151)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at com.keuin.kbackupfabric.util.ZipUtil.makeBackupZip(ZipUtil.java:160)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at com.keuin.kbackupfabric.operation.backup.method.ConfiguredPrimitiveBackupMethod.backup(ConfiguredPrimitiveBackupMethod.java:49)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at com.keuin.kbackupfabric.operation.backup.method.ConfiguredPrimitiveBackupMethod.backup(ConfiguredPrimitiveBackupMethod.java:19)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at com.keuin.kbackupfabric.operation.BackupOperation.async(BackupOperation.java:47)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at com.keuin.kbackupfabric.operation.abstracts.InvokableAsyncBlockingOperation$1.async(InvokableAsyncBlockingOperation.java:15)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at com.keuin.kbackupfabric.operation.abstracts.InvokableAsyncOperation$1.async(InvokableAsyncOperation.java:11)
[18:08:37] [BackupWorker/INFO]: [STDERR]: 	at java.base/java.lang.Thread.run(Thread.java:829)
[18:09:08] [Server thread/WARN]: Failed to save player data for Zedseven
[18:09:08] [Server thread/ERROR]: Couldn't save stats
java.io.IOException: No space left on device
	at java.io.FileOutputStream.writeBytes(Native Method) ~[?:?]
	at java.io.FileOutputStream.write(FileOutputStream.java:341) ~[?:?]
	at org.apache.commons.io.IOUtils.write(IOUtils.java:1857) ~[intermediary-server.jar:?]
	at org.apache.commons.io.FileUtils.writeStringToFile(FileUtils.java:1982) ~[intermediary-server.jar:?]
	at org.apache.commons.io.FileUtils.writeStringToFile(FileUtils.java:2017) ~[intermediary-server.jar:?]
	at net.minecraft.class_3442.method_14912(class_3442.java:62) ~[intermediary-server.jar:?]
	at net.minecraft.class_3324.method_14577(class_3324.java:353) ~[intermediary-server.jar:?]
	at net.minecraft.class_3324.method_14617(class_3324.java:694) ~[intermediary-server.jar:?]
	at n
```

Luckily I should be able to recover that data due to my previous backups, but it would be nice if KB handled the issue more gracefully and avoided player data corruption.

Thank you for the great mod!

**Comment by zedseven** [comment]
Looks like it also corrupted my statistics data, but luckily I was able to restore the files from backup for both advancements and stats.

**Comment by keuin** [comment]
I feel sorry for that. It looks like that the disk is filled up when KBackup is making a new `.zip` backup and failed. Then Minecraft tried to save the world and it failed.
On principle this is a Minecraft bug, because Minecraft did not handle the IOException when saving itself. However, KBackup could mitigate this by removing incomplete `.zip` file and print an error in the log. This is a severe problem and I will fix it soon.

**Comment by keuin** [comment]
I checked the source code and it seemed that version 1.4.4 had fixed this bug. Which version of KBackup are you using? If I am correct, upgrading to the latest version of KBackup could solve your problem. The [latest version](https://github.com/keuin/KBackup-Fabric/releases/tag/v1.6.3) just works fine with Minecraft 1.14.x~1.16.x.

**Comment by zedseven** [comment]
No worries, no harm done. I'm running KBackup 1.6.3:

![image](https://user-images.githubusercontent.com/25164338/121630001-d80d6500-ca6b-11eb-9545-4356b057dc5e.png)

(including the full modlist, though I doubt it was due to a mod incompatibility)

**Comment by keuin** [comment]
Thanks for your report. This bug is fixed in [Version 1.6.5](https://github.com/keuin/KBackup-Fabric/releases/tag/v1.6.5).

**Comment by zedseven** [comment]
Thanks for your quick response and patch!
